### DISCUSSION TITLE
Yarn constraints for the common man/woman

### DISCUSSION THREAD

**Discussion by srmagura** [yarnpkg/berry#3468] [Ideas]
First off, I want to make it clear that this is a question / suggestion and not a complaint. I love how Yarn is constantly innovating.

I learned about [Constraints](https://yarnpkg.com/features/constraints) yesterday and my initial reaction was "wow this is super complicated and I have no idea how Prolog works". The Prolog-based constraints system seems powerful and elegant but very hard to approach for the average JavaScript developer. 

I'm sure I could learn Prolog given enough time, but ideally you shouldn't have to learn a new programming language to use Yarn constraints (especially when that programming language is **very** different from traditional programming languages.)

**I was wondering if Yarn constraints could be configured using a simpler and more familiar syntax.** Yarn could either handle the simple syntax directly, or compile the simple syntax to the more powerful Prolog syntax. Yarn could continue to support the Prolog syntax for people who need really complex constraints, but the idea is that the majority of developers would use the simple syntax.

### Some options for the simple syntax

Yarn could support **ONE** of the following formats for configuring constraints:

- JavaScript (CommonJS module)
- JSON
- YAML
- a custom format (you could call it a Domain-Specific Language)

### Common use cases

The simple syntax should support the all common use cases but would not need to support every use case. Here are some guesses at common use cases for constraints. (Using TypeScript, jQuery, and React as examples — it could be any package.)

- Ensure there are no conflicting dependencies across all packages. (This constraint is not met if package1 depends on `typescript@~4.0.0` and package2 depends on `typescript@^4.4.3`.)
- Ensure all packages depend on the same version of `typescript`. (This constraint is not met if package1 depends on `typescript@^4.0.0` and package2 depends on `typescript@^4.1.0` and Yarn resolves these semver ranges to different versions.)
- Ensure all packages that depend on `typescript` use version >= 4.0.0.
- Ensure all packages that depend on `typescript` use version >= 4.0.0 and < 5.0.0.
- Ensure all packages that depend on `@react-navigation/*` use version >= 6.0.0. (Using the wildcard character to refer to multiple packages with similar names.)
- Ensure no package depends on `jquery`.
- Ensure all peer dependencies are satisfied.
- Ensure all packages that depend on `react` also depend on `react-dom`.
- Ensure that `typescript`, if referenced, is in `devDependencies` and not in `dependencies`.
- Ensure all workspace dependencies use `workspace:`.

...and probably more that I haven't thought of.

_Idea:_ Each of these bullet points could be thought of as a "rule", similar to a rule in ESLint.

### Proposal for simple syntax
The question is, what would this simple syntax actually look like? I think that's outside the scope of this post but I'm happy to come up with something if this idea gains traction.

Lastly, feel free to ignore this suggestion if it doesn't make sense. 😂

**Comment by arcanis** [comment]
We had some discussion about providing a TS API that would "transpile" to Prolog: https://github.com/yarnpkg/berry/issues/1276. No concrete work though.
